tanner/session_analyzer.py

```diff
@@ -114,26 +112,14 @@ def choose_possible_owner(stats):
                       'Mozilla/5.0 (Windows Phone 8.1; ARM; Trident/7.0; Touch; rv:11.0; '
                       'IEMobile/11.0; NOKIA; Lumia 530) like Gecko (compatible; bingbot/2.0; '
                       '+http://www.bing.com/bingbot.htm)']
-        if stats['user_agent'] in bots_owner:
-            hostname, _, _ = socket.gethostbyaddr(stats['peer_ip'])
-            if 'search.msn.com' or 'googlebot.com' in hostname:
-                possible_owners['crawler'] += 1
-            else:
-                possible_owners['attacker'] += 1
-        if stats['requests_in_second'] >= 10:
-            possible_owners['tool'] += 1
-            possible_owners['crawler'] += 1
-        else:
-            possible_owners['user'] += 1
-            possible_owners['attacker'] += 1
-        if stats['hidden_links'] > 0:
-            possible_owners['crawler'] += 1
-            possible_owners['attacker'] += 1
-        if set(stats['attack_types']).intersection(attacks):
-            possible_owners['attacker'] += 1
+        possible_owners['crawler'], possible_owners['tool'] = await self.detect_crawler(stats, bots_owner)
+        possible_owners['attacker'] = await self.detect_attacker(stats, bots_owner, attacks)
+
+        maxcf = max([possible_owners['crawler'], possible_owners['attacker'], possible_owners['tool']])
 
-        maxval = max(possible_owners.items(), key=operator.itemgetter(1))[1]
-        owners = [k for k, v in possible_owners.items() if v == maxval]
+        possible_owners['user'] = round(1 - maxcf, 2)
```

Comment:
could you please explain the purpose of round? What is wrong with simple subtraction?

Reply:
This is because, In other owner cases we assign the cf values ourselves with 2 decimal points. Where as in this case, we do a floating point subtraction, which can (or might) lead to several decimal points, sometimes it might even contain dummy values in the end. 
I found this in a case where i expected cf-user to be `0.15`, but the value came out to be `0.15000000000000002`. This is why I used `round` function to round off the values to 2 decimal points, to avoid any errors or confusions.